cli/src/main.rs

```diff
@@ -0,0 +1,375 @@
+    #[clap(long)]
+    threshold: u64,
+
+    /// The public keys of the multisig owners, who can sign transactions.
+    #[clap(long = "owner")]
+    owners: Vec<Pubkey>,
+}
+
+#[derive(Clap, Debug)]
+struct ProposeUpgradeOpts {
+    /// The multisig account whose owners should vote for this proposal.
+    #[clap(long)]
+    multisig_address: Pubkey,
+
+    /// The program id of the program to upgrade.
+    #[clap(long)]
+    program_address: Pubkey,
+
+    /// The address that holds the new program data.
+    #[clap(long)]
+    buffer_address: Pubkey,
+
+    /// Account that will receive leftover funds from the buffer account.
+    #[clap(long)]
+    spill_address: Pubkey,
+}
+
+#[derive(Clap, Debug)]
+struct ShowMultisigOpts {
+    /// The multisig account to display.
+    #[clap(long)]
+    multisig_address: Pubkey,
+}
+
+#[derive(Clap, Debug)]
+struct ShowTransactionOpts {
+    /// The transaction to display.
+    #[clap(long)]
+    transaction_address: Pubkey,
+}
+
+#[derive(Clap, Debug)]
+struct ApproveOpts {
+    /// The multisig account whose owners should vote for this proposal.
+    // TODO: Can be omitted, we can obtain it from the transaction account.
+    #[clap(long)]
+    multisig_address: Pubkey,
+
+    /// The transaction to approve.
+    #[clap(long)]
+    transaction_address: Pubkey,
+}
+
+/// Resolve ~/.config/solana/id.json.
+fn get_default_keypair_path() -> PathBuf {
+    let home = std::env::var("HOME").expect("Expected $HOME to be set.");
+    let mut path = PathBuf::from(home);
+    path.push(".config/solana/id.json");
+    path
+}
+
+fn main() {
+    let opts = Opts::parse();
+    let payer_keypair_path = match opts.keypair_path {
+        Some(path) => path,
+        None => get_default_keypair_path(),
+    };
+    let payer = read_keypair_file(&payer_keypair_path)
+        .expect(&format!("Failed to read key pair from {:?}.", payer_keypair_path));
+
+    let client = Client::new_with_options(
+        Cluster::Localnet,
+        payer,
+        CommitmentConfig::confirmed(),
+    );
+    let program = client.program(opts.multisig_program_id);
+
+    match opts.subcommand {
+        SubCommand::CreateMultisig(cmd_opts) => create_multisig(program, cmd_opts),
+        SubCommand::ShowMultisig(cmd_opts) => show_multisig(program, cmd_opts),
+        SubCommand::ShowTransaction(cmd_opts) => show_transaction(program, cmd_opts),
+        SubCommand::ProposeUpgrade(cmd_opts) => propose_upgrade(program, cmd_opts),
+        SubCommand::Approve(cmd_opts) => approve(program, cmd_opts),
+    }
+}
+
+fn get_multisig_program_address(
+    program: &Program,
+    multisig_pubkey: &Pubkey,
+) -> (Pubkey, u8) {
+    let seeds = [
+        multisig_pubkey.as_ref(),
+    ];
+    Pubkey::find_program_address(
+        &seeds,
+        &program.id(),
+    )
+}
+
+fn create_multisig(program: Program, opts: CreateMultisigOpts) {
+    if opts.threshold > opts.owners.len() as u64 {
```

Comment:
Isn't `threshold` the *minimum amount of owners required*?  which would make this check weird

Reply:
Yes, it is the minimum amount of owners whose signatures are required to execute the transaction. Therefore when creating the multisig, it makes no sense to specify a higher threshold than the number of owners, because then the multisig could never execute anything. This check guards against that.